File: src/GameData/state_machine.cpp
#include <sengine/state_machine.h>

namespace SEngine
{
    State::State() {}

    void State::setup(Logger *logger) {
        this->log = logger;
    }
    
    void State::update(float dt) {}

    void State::draw(Context *ctx) {}

    State* State::get_state() {
        return static_cast<State*>(this);
    }

    StateMachine::StateMachine(Logger *logger) {
        this->current_state = nullptr;
        this->log = logger;
    }

    StateMachine::StateMachine() {
        this->current_state = nullptr;
    }

    void StateMachine::set_state(State *state) {
        this->current_state = state;
        this->current_state->setup(this->log);
    }

    void StateMachine::update(float dt) {
        this->current_state->update(dt);
    }

    void StateMachine::draw(Context *ctx) {
        this->current_state->draw(ctx);
    }

    StateMachine::~StateMachine() {}
} // namespace SEngine

File: src/data/state_machine.rs
use crate::render::context::Context;

pub trait State {
    fn update(&mut self, dt: f32);
    fn draw(&self, ctx: &mut Context);
}

pub struct StateMachine<S>
where
    S: State,
{
    state: Option<S>,
}

impl<S> Default for StateMachine<S>
where
    S: State,
{
    fn default() -> Self {
        Self { state: None }
    }
}

impl<S> StateMachine<S>
where
    S: State,
{
    pub fn set_state(&mut self, state: S) {
        self.state = Some(state);
    }

    pub fn update(&mut self, dt: f32) {
        self.state.as_mut().unwrap().update(dt);
    }

    pub fn draw(&mut self, ctx: &mut Context) {
        self.state.as_mut().unwrap().draw(ctx);
    }
}
